Feat/level3

nakatomoyanyan opened the pull request:
⚪︎予想時間
12h
⚪︎実装時間
13h8m
⚪︎変更点の概要
・json-server を使ってモックサーバーを作った
・DBの構造は現状のタスク管理アプリにあわせた（タスク名・優先度・期限 + id　を加える）
・事前に数件のデータをモックサーバーのDBに作った
・画面が表示される段階で、登録済みのタスクを取得して表示させた
・APIからデータを取得する処理は別の JS ファイルに書いた
・非同期処理を実装した
・セレクトボックスから優先度の高い順、低い順のどちらかを選び、並び替えられるようにした
・同じ優先度内では期限の近い順に並び替えた
・並び替えにはLodashのorderByを使用する
⚪︎実装した機能の指示内容
https://qiita.com/foolproof-htb/private/7389bc7125fed9635c8e
⚪︎レイアウトの画像
<img width="1487" alt="img_0725_1458" src="https://github.com/user-attachments/assets/dbd181d1-5ac9-437c-8872-63f7fc08a5cd">
⚪︎動作のgif
![level3_1](https://github.com/user-attachments/assets/09a673cc-1d6d-40c5-a596-c68de041a935)

nakatomoyanyan committed:
db.jsonの作成

変更箇所：db.jsonを作成しました。

nakatomoyanyan committed:
モックサーバーに2件のデータを作成

変更内容：モックサーバーに2件のデータを作成しました。
データの構造は「id、タスク名、優先度、期限」です。

nakatomoyanyan committed:
モックサーバーに事前に登録されているタスクを表示

実装内容：モックサーバーからdb.jsonに登録されたタスクの情報を取得。その情報を元にタスクを作成。また情報の取得には非同期処理を実装した。

nakatomoyanyan committed:
作成したタスクのデータをdb.JSONに追加

変更内容：作成したタスクのデータのJSONデータをfetchAPIを用いてdb.JSONに追加。今回の実装によりwebアプリを読み込み直してもデータが消えなくなった。

nakatomoyanyan committed:
削除したタスクをdb.jsonからも削除

実装内容：削除したタスクをfetchを用いてdb.jsonからも削除。削除したいタスクを指定するためのidをタスク作成時に付与。

nakatomoyanyan committed:
lodashのインストール

実装内容：lodashを実装環境にインストールした。

nakatomoyanyan committed:
リファクタリング

実装内容：
・追加ボタンのタイプをsubmitからbuttonに変更。目的は追加ボタンクリック時のエラーを解消するため。
・jsonファイルの情報を読みこむ処理を別のjsファイルに記述した。
・追加タスクの情報をモックサーバーに登録後に、jsonファイルを読み込んでタスクを表示。

nakatomoyanyan committed:
タスクを優先度順でソートする

実装内容：タスクを優先度中でlodashを用いてソートする。優先度が同じ場合は期日が近い順に並べる。タスクの優先度の値を数値で表すようにした。

iwaguma-manabi commented on task_add.js:
```diff
@@ -30,32 +32,113 @@ function addRemoveEvent(newTask){
+    `;
+    addCompleteEvent(newTask);
+    addNotCompleteEvent(newTask);
+    addRemoveEvent(newTask);
```
ソートの際にタスクを0から自力で作り直しというのが保守性に懸念があります。

作り直さずにDOMのまま並び替える
または
タスクの作成処理を共通化（createTaskも含めての共通化の意図）

のどちらかは必要かと思いましたがどうでしょう？

foolproof-htb commented on create_task_from_json.js:
```diff
@@ -0,0 +1,62 @@
+let url = 'http://localhost:2000/tasks';
+
+function addCompleteEvent(newTask){
+    let completeButton = newTask.querySelector('.complete');
```
細かい点で恐縮です！
なんかこの行以降のインデントが揃ってない気がします！

SaitoJP commented on task_add.js:
```diff
@@ -1,9 +1,11 @@
+let taskImportance = document.querySelector('.task_importance');
+let taskSubmit = document.querySelector('.task_submit');
 let taskList = document.querySelector('.task_list');
+let sort_by_importance = document.querySelector('.sort_by_importance');
```
細かいですが、再代入しない変数は全て `const` で宣言したほうが良いです。
`const` で宣言すると意図しない再代入はエラーになり抑止できます。
なので、基本的には `const` を使用し、再代入される可能性があれば `let` を使うイメージ。

SaitoJP commented on create_task_from_json.js:
```diff
@@ -0,0 +1,62 @@
+let taskList = document.querySelector('.task_list');
+let url = 'http://localhost:2000/tasks';
```
定数は変数名を大文字にして const で定義するのが良いですよ。
`const URL = 'http://localhost:2000/tasks'`

SaitoJP commented on create_task_from_json.js:
```diff
@@ -0,0 +1,62 @@
+    let result = await fetch(deleteTaskURL,{
+      method: "DELETE",
+    })
+  };
```
APIリクエストをまとめた api.js ファイルを作成するのはどうでしょう。
リクエスト先（エンドポイント）の定義を一箇所にまとめられてコードの見通しが良くなります。

api.js
```js
const ENDPOINT_URL =  'http://localhost:2000/tasks'

export async function removeTaskApi(taskId) {
    const url = `${ENDPOINT_URL}/${taskId}`;
    const result = await fetch(url, { method: "DELETE", });

    // エラーハンドリングをする
    if (!result.ok) {
        throw new Error(`HTTP error! status: ${result.status}`);
    }
    
   return;
}


// 以下タスク追加処理など、fetch処理部分は本ファイルにまとめる想定
```

nakatomoyanyan committed:
インデントの修正

変更内容：関数内のインデントが揃っていなかったため、インデントのズレを修正した。

nakatomoyanyan committed:
定数をconstを用いて定義する

変更内容：今までletを用いて定義していた定数をconstを用いて定義し直した。

nakatomoyanyan committed:
api.jsの作成

変更内容：APIリクエストをまとめたapi.jsを作成した。create_task_from_json.jsは削除した。またエラーハンドリングを実装した。

nakatomoyanyan committed:
タスクのソート機能のリファクタリング

変更内容：タスクを作り直さずにDOMのまま並べ替えるように変更
変更理由：ソートの際にタスクを０から作り直す方法には保守性に懸念があったため。
実装方法：idでタスクを指定後、removechildとappendchildを使用してDOMのまま並べ替える。

nakatomoyanyan committed:
prettierの再インストール

藤澤さんの指示の元、prettierを再インストールしました。

nakatomoyanyan committed:
API操作の関数とDOM操作の関数を別ファイルに分ける

変更内容：API操作の関数をapi.jsに、DOM操作の関数をdom.jsに分けた。

nakatomoyanyan committed:
prettierによる整形

変更内容：prettierによるコードの整形
